src/samples/permissionhooks/AllowlistPlugin.sol

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.25;
+
+import {PackedUserOperation} from "@eth-infinitism/account-abstraction/interfaces/PackedUserOperation.sol";
+
+import {PluginMetadata, PluginManifest} from "../../interfaces/IPlugin.sol";
+import {IValidationHook} from "../../interfaces/IValidationHook.sol";
+import {IStandardExecutor, Call} from "../../interfaces/IStandardExecutor.sol";
+import {BasePlugin} from "../../plugins/BasePlugin.sol";
+
+contract AllowlistPlugin is IValidationHook, BasePlugin {
+    enum FunctionId {
+        PRE_VALIDATION_HOOK
+    }
+
+    struct AllowlistInit {
+        address target;
+        bool hasSelectorAllowlist;
+        bytes4[] selectors;
+    }
+
+    struct AllowlistEntry {
+        bool allowed;
+        bool hasSelectorAllowlist;
+    }
+
+    mapping(address target => mapping(address account => AllowlistEntry)) public targetAllowlist;
+    mapping(address target => mapping(bytes4 selector => mapping(address account => bool))) public
+        selectorAllowlist;
+
+    error TargetNotAllowed();
+    error SelectorNotAllowed();
+    error NoSelectorSpecified();
+
+    function onInstall(bytes calldata data) external override {
+        AllowlistInit[] memory init = abi.decode(data, (AllowlistInit[]));
+
+        for (uint256 i = 0; i < init.length; i++) {
+            targetAllowlist[init[i].target][msg.sender] = AllowlistEntry(true, init[i].hasSelectorAllowlist);
+
+            if (init[i].hasSelectorAllowlist) {
+                for (uint256 j = 0; j < init[i].selectors.length; j++) {
+                    selectorAllowlist[init[i].target][init[i].selectors[j]][msg.sender] = true;
+                }
+            }
+        }
+    }
+
+    function onUninstall(bytes calldata data) external override {
+        AllowlistInit[] memory init = abi.decode(data, (AllowlistInit[]));
+
+        for (uint256 i = 0; i < init.length; i++) {
+            delete targetAllowlist[init[i].target][msg.sender];
+
+            if (init[i].hasSelectorAllowlist) {
+                for (uint256 j = 0; j < init[i].selectors.length; j++) {
+                    delete selectorAllowlist[init[i].target][init[i].selectors[j]][msg.sender];
+                }
+            }
+        }
+    }
+
+    function setAllowlistTarget(address target, bool allowed, bool hasSelectorAllowlist) external {
+        targetAllowlist[target][msg.sender] = AllowlistEntry(allowed, hasSelectorAllowlist);
+    }
+
+    function setAllowlistSelector(address target, bytes4 selector, bool allowed) external {
+        selectorAllowlist[target][selector][msg.sender] = allowed;
+    }
+
+    function preUserOpValidationHook(uint8 functionId, PackedUserOperation calldata userOp, bytes32)
+        external
+        view
+        override
+        returns (uint256)
+    {
+        if (functionId == uint8(FunctionId.PRE_VALIDATION_HOOK)) {
```

Comment:
What's the point of this check, and the similar one in the runtime validation hook?

Reply:
In other plugin examples we've been using an if-else chain for the internal function dispatcher, so I was following that pattern here too. I suppose we could cut it and just treat every function id the same here, because this one doesn't have multiple implementations.

Reply:
Like the idea of not having the functionId, especially the plugins are global singleton.